plugins/inputs/dpdk/dpdk_connector.go

```diff
@@ -0,0 +1,145 @@
+// +build linux
+
+package dpdk
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+)
+
+const maxInitMessageLength = 1024
+
+type initMessage struct {
+	Version      string `json:"version"`
+	Pid          int    `json:"pid"`
+	MaxOutputLen uint32 `json:"max_output_len"`
+}
+
+type dpdkConnector struct {
+	pathToSocket  string
+	maxOutputLen  uint32
+	messageShowed bool
+	accessTimeout time.Duration
+	connection    net.Conn
+}
+
+func newDpdkConnector(pathToSocket string, accessTimeout config.Duration) *dpdkConnector {
+	return &dpdkConnector{
+		pathToSocket:  pathToSocket,
+		messageShowed: false,
+		accessTimeout: time.Duration(accessTimeout),
+	}
+}
+
+// Connects to the socket
+// Since DPDK is a local unix socket, it is instantly returns error or connection, so there's no need to set timeout for it
+func (conn *dpdkConnector) connect() (string, error) {
+	connection, err := net.Dial("unixpacket", conn.pathToSocket)
+
+	if err != nil {
+		return "", fmt.Errorf("failed to connect to the socket - %v", err)
+	}
+
+	result, err := conn.readMaxOutputLen(connection)
+	if err != nil {
+		return result, err
+	}
+
+	conn.connection = connection
+	return result, nil
+}
+
+// Executes command using provided connection and returns response
+// If error (such as timeout) occurred, then connection is discarded and recreated
+// because otherwise behaviour of connection is undefined (e.g. it could return result of timed out command instead of latest)
+func (conn *dpdkConnector) getCommandResponse(fullCommand string) ([]byte, error) {
+	connection, err := conn.getConnection()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connection to execute %v command - %v", fullCommand, err)
+	}
+
+	err = conn.setTimeout(connection)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set timeout for %v command - %v", fullCommand, err)
+	}
+
+	_, err = connection.Write([]byte(fullCommand))
+	if err != nil {
+		return nil, conn.tryClose(fmt.Sprintf("send '%v'", fullCommand), err)
+	}
+
+	buf := make([]byte, conn.maxOutputLen)
+	messageLength, err := connection.Read(buf)
+	if err != nil {
+		return nil, conn.tryClose(fmt.Sprintf("read response of '%v'", fullCommand), err)
+	}
+
+	if messageLength == 0 {
+		return nil, fmt.Errorf("got empty response during execution of '%v' command", fullCommand)
+	}
+	return buf[:messageLength], nil
+}
+
+func (conn *dpdkConnector) tryClose(operation string, err error) error {
+	closeErr := conn.connection.Close()
```

Comment:
`conn.connection` might be `nil` in case `connect()` was never called (successfully).

Reply:
In case `connect()` was never successfully called, `tryClose` function will never be called.

Reply:
That's maybe the current state of afairs, but please also protect our children and grad-children from shooting themselves in the foot! ;-)

Reply:
I would definitely want to avoid checking in every unexported function if argument is `nil`. 
It would be different case for exported functions.

Reply:
I would definitely want to avoid checking in every unexported function if argument is `nil`.
It would be different case for exported functions.

Reply:
Really? The one `if` I'm requesting? Handling the connection state is often a difficult thing so I think an extra guard cannot hurt...

Reply:
@srebhan 
Ok, I will do that but since you want to check if connection was established (is not nil) in `dpdk_connector.go`, I will not make any changes for your comment regarding checking status of connection in `Gather(...)`. I would like to have connection-related activities in this file.

Reply:
Ok I can live with this.

Reply:
Good to hear :)